### ISSUE TITLE
std::equal gives wrong answer on array of arrays

### ISSUE THREAD

**Issue by AMP999** [llvm/llvm-project#63371]
https://godbolt.org/z/fcchWzz9n
```
    int a[2][2] = { {1,1}, {2,2} };
    int b[2][2] = { {1,1}, {2,2} };
    assert(std::ranges::equal(a, b));
      // libc++: wrongly succeeds
      // libstdc++, MSVC: correctly fails the assertion
```
libc++'s `std::ranges::equal` is optimized for element types that are `__is_trivially_equality_comparable`.
This optimization should be disabled when the element type is an array type, because arrays are comparable only by array-to-pointer conversion. So direct comparison of array types is non-trivial:
```
    assert(memcmp(&a[0], &b[0], sizeof a[0]) == 0);
    assert(a[0] != b[0]);
```
Alternatively, this bug could be fixed by changing `QualType::isTriviallyEqualityComparableType` to reject array types. But if we did that, then we would change the triviality of `S` in this example:
```
    using A = int[2];
    static_assert(__is_trivially_equality_comparable(A));
      // True today, but false in the hypothetical: No problem
    struct S { A a; bool operator==(const S&) const = default; };
    static_assert(__is_trivially_equality_comparable(S));
      // Correctly true today, but false in the hypothetical: Yuck!
```
Therefore I think this should be fixed on the libc++ side (to disable the optimization specifically for array types), and not on the Clang side.

**Comment by philnik777** [comment]
I agree that this is a bug, but I think this should be fixed in Clang. We can just check whether the top-level type is an array and return early. `isTriviallyEqualityComparableType` isn't called recursively, so a type containing an array would still work just fine.
